crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1516,6 +1516,96 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    fn infer_augmented_op(
+        &mut self,
+        assignment: &ast::StmtAugAssign,
+        target_type: Type<'db>,
+        value_type: Type<'db>,
+    ) -> Type<'db> {
+        // If the target defines, e.g., `__iadd__`, infer the augmented assignment as a call to that
+        // dunder.
+        let op = assignment.op;
+        match target_type {
+            Type::Union(union) => {
+                return union.map(self.db, |&target_type| {
+                    self.infer_augmented_op(assignment, target_type, value_type)
+                })
+            }
+            Type::Instance(class) => {
+                if let Symbol::Type(class_member, boundness) =
+                    class.class_member(self.db, op.in_place_dunder())
+                {
+                    let call = class_member.call(self.db, &[target_type, value_type]);
```

Comment:
I feel like there may still be a missing case here... Let's see if I can articulate it after his meeting.

Reply:
```python
def bool_instance() -> bool:
    return True

flag = bool_instance()

class Foo:
    def __add__(self, other: int) -> str:
        return "Hello, world!"
    if bool_instance():
        def __iadd__(self, other: int) -> int:
            return 42


class Bar:
    def __add__(self, other: int) -> bool:
        return True

    def __iadd__(self, other: int) -> float:
        return 42.0
        
if flag:
    f = Foo()
else:
    f = Bar()
f += 12
```

I was thinking of something like this, where we don't know _which_ part of a union was unbound, and so we incorrectly include the `__add__` types for the _bound_ part of the union too... But this case is actually handled correctly right now.

Reply:
Worth adding a test for this, if you weren't sure?

Reply:
I did! `## Partially bound target union with __add__`